### PR TITLE
refactor: add no-return-await lint rule

### PR DESCRIPTION
**Pull request by bartlomieju**
https://eslint.org/docs/rules/no-return-await

### CODE REVIEW

**Comment by bartlomieju** [comment]
> LGTM, awesome...
> 
> When await is not used in the method/function block, does `async` cause another microtask to be scheduled?

Not sure, but thanks for the lead.

> Either way, if you are returning a promise in all branches (like a lot of these updates) the method no longer needs to be `async`. There is an eslint rule of [`require-await`](https://eslint.org/docs/rules/require-await) which would catch this as well. With TypeScript checking the return type it is very unlikely enforcing this would cause incorrect behaviour.

I'll see to it tomorrow

**Comment by nayeemrmn** [comment]
> Either way, if you are returning a promise in all branches (like a lot of these updates) the method no longer needs to be `async`

Hmm, isn't it a good indicator for a function that returns a Promise to be marked async? Or does it affect performance? (I guess I disagree with "either way".)

**Comment by kitsonk** [comment]
> Hmm, isn't it a good indicator for a function that returns a Promise to be marked async?

It is an internal concern, it has no external detectable difference...  When you generate a type definition for example, there is no indication that the method is async or not.

File: core/examples/http_bench.js [hunk 1]
```diff
@@ -85,20 +85,20 @@ function listen() {
...
  */
 async function read(rid, data) {
-  return await sendAsync(ops["read"], rid, data);
+  return sendAsync(ops["read"], rid, data);
 }
 
 /** Writes a fixed HTTP response to the socket rid. Returns bytes written. */
 async function write(rid, data) {
-  return await sendAsync(ops["write"], rid, data);
+  return sendAsync(ops["write"], rid, data);
```
Comment (ry) [inline core/examples/http_bench.js]: It will be interesting to see if this has any effect.

Comment (kitsonk) [reply core/examples/http_bench.js]: It should.  It is at least one less microtask, unless V8 optimises it out, which it might be smart enough to do.

**Comment by bartlomieju** [comment]
> There is an eslint rule of [`require-await`](https://eslint.org/docs/rules/require-await) which would catch this as well. With TypeScript checking the return type it is very unlikely enforcing this would cause incorrect behaviour.

@kitsonk I did a quick check and we'd have to add a lot of supressions to tests (because there's a lot of mocks that assert async functions were called). I'm open to that change, but I'd open a new issue for that.

**Pull request merged**